### PR TITLE
ci: update to latest intel osx runner and download 10.15 SDK

### PR DESCRIPTION
**Pull request by beckermr**
### Description

This PR updates the CI to the latest intel osx runner and to download the 10.15 SDK.

I have preserved the older option of directly using the system SDK via xcode. This should make it easier for folks to work locally in some cases without doing their own SDK download.

Resolves https://github.com/conda/conda-build/issues/5378

### Checklist - did you ...

- [x] Add a file to the `news` directory ([using the template](../blob/main/news/TEMPLATE)) for the next release's release notes?
     
- [x] Add / update necessary tests?
- [x] Add / update outdated documentation?

### CODE REVIEW

File: .github/workflows/tests.yml [hunk 1]
```diff
@@ -415,6 +408,8 @@ jobs:
           --cov=conda_build
           -n auto
           -m "${{ env.PYTEST_MARKER }}"
+        env:
+          CONDA_SUBDIR: osx-64
```
Comment (beckermr) [inline .github/workflows/tests.yml:412]:
```suggestion
```

**Comment by kenodegard** [comment]
I also started on this in https://github.com/conda/conda-build/issues/5379

While the SDK downloads are working I haven't quite gotten the tests to pass again

File: .github/workflows/tests.yml [hunk 2]
```diff
@@ -337,11 +337,7 @@ jobs:
     needs: changes
     if: github.event_name == 'schedule' || needs.changes.outputs.code == 'true'
 
-    # Old macOS needed for old SDK (see xcode step below)
-    # This is needed for some MACOSX_DEPLOYMENT_TARGET tests
-    # We could also install SDKs from a external provider in the future
-    # if we want to update this runner to a non-deprecated version
-    runs-on: macos-11
+    runs-on: macos-latest
```
Comment (beckermr) [inline .github/workflows/tests.yml:340]:
```suggestion
    # we still need intel macs so we are stuck on macos-13 (not -14 or -latest)
    # the issue is that there are recipes that depend on packages
    # that do not exist for osx-arm64 - see #5388
    runs-on: macos-13
```

File: .github/workflows/tests.yml [hunk 3]
```diff
@@ -385,9 +384,6 @@ jobs:
           condarc-file: .github/condarc
           run-post: false  # skip post cleanup
```
Comment (beckermr) [inline .github/workflows/tests.yml:386]:
```suggestion
      # this step installs the 13.3 SDK
      # see https://github.com/actions/runner-images/blob/main/images/macos/macos-13-Readme.md
      - name: Xcode Install
        run: sudo xcode-select --switch /Applications/Xcode_14.3.1.app
```

File: .github/workflows/tests.yml [hunk 4]
```diff
@@ -384,10 +383,10 @@ jobs:
         with:
           condarc-file: .github/condarc
           run-post: false  # skip post cleanup
-
+      # this step installs the 13.3 SDK
+      # see https://github.com/actions/runner-images/blob/main/images/macos/macos-13-Readme.md
       - name: Xcode Install
-        run: sudo xcode-select --switch /Applications/Xcode_11.7.app
-
+        run: sudo xcode-select --switch /Applications/Xcode_14.3.1.app
```
Comment (beckermr) [inline .github/workflows/tests.yml:389]:
```suggestion
        run: sudo xcode-select --switch /Applications/Xcode_14.3.1.app

```

File: .github/workflows/tests.yml [hunk 5]
```diff
@@ -384,9 +383,10 @@ jobs:
         with:
           condarc-file: .github/condarc
           run-post: false  # skip post cleanup
-
+      # this step installs the 13.3 SDK
```
Comment (beckermr) [inline .github/workflows/tests.yml:386]:
```suggestion

      # this step installs the 13.3 SDK
```

File: .github/workflows/tests.yml [hunk 6]
```diff
@@ -385,8 +386,28 @@ jobs:
...
+          export SDKROOT=${SDK_DIR}/MacOSX${MACOSX_SDK_VERSION}.sdk
+
+          # sudo xcode-select --switch /Applications/Xcode_14.3.1.app
+
+          if [ ! -d ${SDK_DIR} ]; then mkdir ${SDK_DIR}; fi
+          if [ ! -d ${SDKROOT} ]; then
+            url="https://github.com/phracker/MacOSX-SDKs/releases/download/11.3/MacOSX${MACOSX_SDK_VERSION}.sdk.tar.xz"
+            curl -L --output MacOSX${MACOSX_SDK_VERSION}.sdk.tar.xz "${url}"
+            sdk_sha256=ac75d9e0eb619881f5aa6240689fce862dcb8e123f710032b7409ff5f4c3d18b
+            echo "${sdk_sha256} *MacOSX11.3.sdk.tar.bz2" | shasum -a 256 -c
```
Comment (beckermr) [inline .github/workflows/tests.yml:405]:
```suggestion
```

File: tests/conftest.py [hunk 7]
```diff
@@ -205,22 +205,11 @@ def variants_conda_build_sysroot(monkeypatch, request):
...
     )
     monkeypatch.setenv(
         "MACOSX_DEPLOYMENT_TARGET",
-        subprocess.run(
-            ["xcrun", "--sdk", "macosx", "--show-sdk-version"],
-            check=True,
-            capture_output=True,
-            text=True,
-        ).stdout.strip(),
+        "11.3",
```
Comment (beckermr) [inline tests/conftest.py:212]:
```suggestion
        os.environ["MACOSX_SDK_VERSION"],
```

File: tests/conftest.py [hunk 8]
```diff
@@ -205,22 +205,11 @@ def variants_conda_build_sysroot(monkeypatch, request):
...
 
     monkeypatch.setenv(
-        "CONDA_BUILD_SYSROOT",
-        subprocess.run(
-            ["xcrun", "--sdk", "macosx", "--show-sdk-path"],
-            check=True,
-            capture_output=True,
-            text=True,
-        ).stdout.strip(),
+        "CONDA_BUILD_SYSROOT", os.path.expanduser("~/macosx_sdks/MacOSX11.3.sdk")
```
Comment (beckermr) [inline tests/conftest.py:208]:
```suggestion
        "CONDA_BUILD_SYSROOT", os.environ["SDKROOT"])
```

**Comment by beckermr** [comment]
The issue is that the test suite appears to be ignoring the downloaded sysroot. I think this must be in how env vars are passed to conda-build executions.

File: news/5387-ci-osx-sdk [hunk 9]
```diff
@@ -0,0 +1,3 @@
+### Other
+
+* Updated the CI to use the OSX 13.3 SDK
```
Comment (beckermr) [inline news/5387-ci-osx-sdk:3]:
```suggestion
* Updated the CI to download the MacOSX 10.15 SDK.
```

File: tests/cli/test_main_build.py [hunk 10]
```diff
@@ -320,11 +320,11 @@ def test_no_force_upload(
 
 @pytest.mark.slow
 def test_conda_py_no_period(testing_workdir, testing_metadata, monkeypatch):
-    monkeypatch.setenv("CONDA_PY", "36")
+    monkeypatch.setenv("CONDA_PY", "311")
```
Comment (beckermr) [inline tests/cli/test_main_build.py:323]:
```suggestion
    monkeypatch.setenv("CONDA_PY", "36")
```

File: tests/cli/test_main_build.py [hunk 11]
```diff
@@ -320,11 +320,11 @@ def test_no_force_upload(
...
     testing_metadata.meta["requirements"] = {"host": ["python"], "run": ["python"]}
     api.output_yaml(testing_metadata, "meta.yaml")
     outputs = api.build(testing_workdir, notest=True)
-    assert any("py36" in output for output in outputs)
+    assert any("py311" in output for output in outputs)
```
Comment (beckermr) [inline tests/cli/test_main_build.py:327]:
```suggestion
    assert any("py36" in output for output in outputs)
```

File: tests/test-recipes/metadata/always_include_files_glob/meta.yaml [hunk 12]
```diff
@@ -23,7 +23,7 @@ requirements:
     # installed.
     - {{ compiler('c') }}  # [linux or osx]
   host:
-    - libpng 1.6.34.*
+    - libpng 1.6.39.*
```
Comment (beckermr) [inline tests/test-recipes/metadata/always_include_files_glob/meta.yaml:26]:
```suggestion
    - libpng 1.6.34.*
```

File: tests/test_api_skeleton.py [hunk 13]
```diff
@@ -387,7 +387,7 @@ def test_list_skeletons():
 
 
 def test_pypi_with_entry_points(tmp_path: Path):
-    api.skeletonize("planemo", repo="pypi", python_version="3.7", output_dir=tmp_path)
+    api.skeletonize("planemo", repo="pypi", python_version="3.11", output_dir=tmp_path)
```
Comment (beckermr) [inline tests/test_api_skeleton.py:390]:
```suggestion
    api.skeletonize("planemo", repo="pypi", python_version="3.7", output_dir=tmp_path)
```

File: tests/test_api_skeleton.py [hunk 14]
```diff
@@ -411,7 +411,7 @@ def test_pypi_with_extra_specs(tmp_path: Path, testing_config):
         "pypi",
         extra_specs=extra_specs,
         version="0.1.24",
-        python="3.6",
+        python="3.11",
```
Comment (beckermr) [inline tests/test_api_skeleton.py:414]:
```suggestion
        python="3.6",
```

File: tests/test_api_skeleton.py [hunk 15]
```diff
@@ -434,7 +434,7 @@ def test_pypi_with_version_inconsistency(tmp_path: Path, testing_config):
         "pypi",
         extra_specs=extra_specs,
         version="0.0.10",
-        python="3.6",
+        python="3.11",
```
Comment (beckermr) [inline tests/test_api_skeleton.py:437]:
```suggestion
        python="3.6",
```

File: .github/workflows/tests.yml [hunk 16]
```diff
@@ -415,6 +431,9 @@ jobs:
           --cov=conda_build
           -n auto
           -m "${{ env.PYTEST_MARKER }}"
+        env:
+          SDKROOT: ${{ env.MACOSX_SDK_ROOT }}
```
Comment (beckermr) [inline .github/workflows/tests.yml:435]:
```suggestion
          # SDKROOT: ${{ env.MACOSX_SDK_ROOT }}
```

File: tests/conftest.py [hunk 17]
```diff
@@ -105,7 +105,19 @@ def boolify(v):
...
         conda_pkg_format=conda_pkg_format_default,
     )
-    result = Config(variant=None, **testing_config_kwargs)
+
+    if on_mac:
+        var_dict = {}
+        if "CONDA_BUILD_SYSROOT" in os.environ:
+            var_dict["CONDA_BUILD_SYSROOT"] = [os.environ["CONDA_BUILD_SYSROOT"]]
+        if "SDKROOT" in os.environ:
+            var_dict["SDKROOT"] = [os.environ["SDKROOT"]]
```
Comment (beckermr) [inline tests/conftest.py:114]:
```suggestion
        # if "SDKROOT" in os.environ:
        #     var_dict["SDKROOT"] = [os.environ["SDKROOT"]]
```

File: tests/conftest.py [hunk 18]
```diff
@@ -105,7 +105,19 @@ def boolify(v):
...
         conda_pkg_format=conda_pkg_format_default,
     )
-    result = Config(variant=None, **testing_config_kwargs)
+
+    if on_mac:
+        var_dict = {}
+        if "CONDA_BUILD_SYSROOT" in os.environ:
+            var_dict["CONDA_BUILD_SYSROOT"] = [os.environ["CONDA_BUILD_SYSROOT"]]
+        # if "SDKROOT" in os.environ:
+        #     var_dict["SDKROOT"] = [os.environ["SDKROOT"]]
```
Comment (beckermr) [inline tests/conftest.py:114]:
```suggestion
```

File: .github/workflows/tests.yml [hunk 19]
```diff
@@ -415,6 +431,9 @@ jobs:
           --cov=conda_build
           -n auto
           -m "${{ env.PYTEST_MARKER }}"
+        env:
+          # SDKROOT: ${{ env.MACOSX_SDK_ROOT }}
```
Comment (beckermr) [inline .github/workflows/tests.yml:435]:
```suggestion
```

File: tests/conftest.py [hunk 20]
```diff
@@ -105,7 +105,17 @@ def boolify(v):
...
-    result = Config(variant=None, **testing_config_kwargs)
+
+    if on_mac:
+        var_dict = {}
+        if "CONDA_BUILD_SYSROOT" in os.environ:
+            var_dict["CONDA_BUILD_SYSROOT"] = [os.environ["CONDA_BUILD_SYSROOT"]]
+        if not var_dict:
+            var_dict = None
+    else:
+        var_dict = None
```
Comment (beckermr) [inline tests/conftest.py:116]:
```suggestion
    if on_mac and "CONDA_BUILD_SYSROOT" in os.environ:
        var_dict = {
            "CONDA_BUILD_SYSROOT": [os.environ["CONDA_BUILD_SYSROOT"]],
        }
    else:
        var_dict = None
```

**Comment by beckermr** [comment]
OK @kenodegard this one is ready for review! 

This PR has the most minimal changes I could make to get the test suite to pass on the macos-13 runners. 

Some notes and comments:

 - Your approach in #5379 for the SDKs is more complete for sure. However, it does couple code in the test suite (the fixture) with the locally cached directory on the runners (`~/macosx_sdks`). 
 - I have opted here to keep all of the SDK handling in GHA and have the test suite use w/e sysroot the user presents to it, either through xcode or the `CONDA_BUILD_SYSROOT` environment variable. 
 - The approach here may help folks who use the test suite locally in that they can easily use w/e sysroot they like for testing by setting the `CONDA_BUILD_SYSROOT`.
 - Intel mac hardware is going away and there are lots of tests that have to be run on osx-64 due to packages not be available on osx-arm64. I documented part of this in issue #5388. We have time to address this one too.

**Comment by beckermr** [comment]
@conda/builds-tools @kenodegard The CI here will break starting TOMORROW if we do not merge something to remove macos-11. Shall we merge this one?

**Review by kenodegard — APPROVED** [review]

**Comment by kenodegard** [comment]
@beckermr thanks for working on this!

**Pull request merged**